File: 1/2021.07.07/706.Design HashMap/python/solution.py
class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.capacity = 1000
        self.data = []
        for _ in range(self.capacity):
            self.data.append([])
    
    def hash(self, key):
        return key % self.capacity
        

    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        h = self.hash(key)
        for i in range(len(self.data[h])):
            node = self.data[h][i]
            if node[0] == key:
                self.data[h][i] = (key, value)
                return
        self.data[h].append((key, value))
        

    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        h = self.hash(key)
        for i in range(len(self.data[h])):
            node = self.data[h][i]
            if node[0] == key:
                return node[1]
        return -1
        

    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        h = self.hash(key)
        for i in range(len(self.data[h])):
            node = self.data[h][i]
            if node[0] == key:
                self.data[h].pop(i)
                return

File: 1/2021.07.07/706.Design HashMap/java/solution.java
class MyHashMap {

    class Node {
        int key;
        int value;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private int capacity = 1000;
    private LinkedList<Node>[] data = new LinkedList[capacity];

    /** Initialize your data structure here. */
    public MyHashMap() {
        for (int i = 0; i < capacity; i++) {
            data[i] = new LinkedList<>();
        }
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        int h = hash(key);
        Iterator<Node> iterator = data[h].iterator();
        while (iterator.hasNext()) {
            Node node = iterator.next();
            if (node.key == key) {
                node.value = value;
                return;
            }
        }
        data[h].add(new Node(key, value));
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        int h = hash(key);
        Iterator<Node> iterator = data[h].iterator();
        while (iterator.hasNext()) {
            Node node = iterator.next();
            if (node.key == key) {
                return node.value;
            }
        }
        return -1;
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
        int h = hash(key);
        Iterator<Node> iterator = data[h].iterator();
        while (iterator.hasNext()) {
            Node node = iterator.next();
            if (node.key == key) {
                data[h].remove(node);
                return;
            }
        }
    }

    private int hash(int key) {
        return key % capacity;
    }
}
